File: baekjoon-online-judge/gold/14503.py
from sys import stdin
input = stdin.readline

def isInArea(x, y) :
    return (0 <= x < N and 0 <= y < M)
    
def clean(x, y, d):
    cnt = 0
    if board[x][y] == 0:
        board[x][y] = -1    #(1)
        cnt += 1
    for _ in range(4) : 
        d = (d + 3) % 4     #(2)
        nx = x + D[d][0] 
        ny = y + D[d][1]
        if isInArea(nx, ny) and board[nx][ny] == 0 :  #(2) - (a)
            return cnt + clean(nx, ny, d)
        #(2) - (b)
    #(2) - (c)
    nx = x - D[d][0]
    ny = y - D[d][1]
    if isInArea(nx, ny) : 
        return cnt + (0 if board[nx][ny] == 1 else clean(nx, ny, d)) #(2) - (d)
        
if __name__ == "__main__" :
    N, M = map(int, input().split())
    D = [(-1, 0), (0, 1), (1, 0), (0, -1)]
    x, y, d = map(int, input().split())
    board = [list(map(int, input().split())) for _ in range(N)]
    print(clean(x, y, d))

File: baekjoon-online-judge/gold/14503.java
import java.io.*;
import java.util.*;

class Main {
	static int N, M;
	static int r, c, d;

	// 0: 북, 1: 동, 2: 남, 3: 서
	static int[] dx = {-1, 0, 1, 0};
	static int[] dy = {0, 1, 0, -1};

	static int[][] board;
	static final int CLEAN = -1, BLOCK = 1, EMPTY = 0;

	private static int simulation() {
		int cleanCnt = 0;

		while (true) {
			if (board[r][c] == EMPTY) {
				board[r][c] = CLEAN;
				++cleanCnt;
			}

			boolean moved = false;
			for (int i = 0; i < 4; ++i) {
				d = (d + 3) % 4;
				int nx = r + dx[d];
				int ny = c + dy[d];

				if (board[nx][ny] == EMPTY) {
					r = nx;
					c = ny;
					moved = true;
					break;
				}
			}

			if (moved) continue;
			int backX = r - dx[d];
			int backY = c - dy[d];

			if (board[backX][backY] == BLOCK) break;
			r = backX;
			c = backY;
		}

		return cleanCnt;
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		// Input
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		st = new StringTokenizer(br.readLine());
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		d = Integer.parseInt(st.nextToken());

		board = new int[N][M];
		for (int i = 0; i < N; i++) {
			board[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		}

		// Solve & Output
		System.out.print(simulation());
	}
}